### PR TITLE
port to targetSdk 29

### LINKED ISSUE #7107: Update targetSdk to 29
The Play store has a hard requirement of targetSdk 29 for app update acceptance beginning November 2020 (which in my experience means Nov 1, not Nov anything else)

I believe we're compatible but we need to do it and check

### PR DESCRIPTION
**Pull request by mikehardy**
## Purpose / Description

This is a bit rough-and-ready, not polished and I had hoped to fix more of the deprecation as I went - but the priority is to know with 100% certainty that we have targetSdk 29 capability as soon as possible in the release-2.14 cycle, before we get too close to the Nov 1, 2020 targetSdk29 deadline

So this makes sure we can do it, even if it is not perfect.

Related deprecation issues may be picked off one by one as we have time, but as most of them are fresh deprecations in API30 it should not be too big a deal.

The biggest item of concern right now (which needs testing prior to merging this) is the full screen / immersive UI action in CompatV30. They *really* changed the way that works, and it may be best to peel that commit out to a separate deprecation issue / PR pair so this can go in regardless

## Fixes
Fixes #7107 

## Approach

Mark deprecation where it was unthinkable to make the change now (AsyncTask etc), add Compat interfaces where quickly possible

## How Has This Been Tested?

So far just API28/29/30 emulators to see the effect, with `./gradlew jacocoTestReport` as the test harness.

Needs a good look all around.

## Learning (optional, can help others)

The WindowInsets stuff is a huge change. The threading stuff (AsyncTask, Handler) is a huge change.
The Preferences API change is really big for us also.

The platform moves quickly

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by mikehardy** [comment]
This is the commit I mentioned that I'm iffy on. It's simply not doing the job right now and I'm not 100% sure how to get it to do the job.

https://github.com/ankidroid/Anki-Android/pull/7112/commits/99aeeb931b06a1932d178dae7ce6dd0a18cc1263

May be best to leave the raw work in place but in comments for now and just mark the (current) CompatV19 version for deprecation with a new issue to complete the work.

**Comment by mikehardy** [comment]
Related to this, @david-allison-1 if we want to do any back-porting of database upgrade warnings like #7072 we need to make sure they are installed in release-2.10 (for API15 people) and release-2.13 (for API16+ people) and released before Nov 1 2020 or the updates will *not* be releasable without this PR. And I do not want to backport this PR.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 1]
```diff
@@ -771,7 +765,7 @@ public void updateNotificationPreference(ListPreference listpref) {
         listpref.setSummary(listpref.getEntry().toString());
     }
 
-    private void updateSummary(Preference pref) {
+    private void updateSummary(android.preference.Preference pref) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:768]: Can we define our own abstract classes, ignore the deprecation there, then use them to get rid of the line noise?

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.java [hunk 2]
```diff
@@ -27,8 +27,10 @@
 
     private CompatHelper() {
 
-        if (getSdkVersion() >= 28) {
-            mCompat = new CompatV28();
+        if (getSdkVersion() >= 30) {
+            mCompat = new CompatV30();
+        } else if (getSdkVersion() >= 28) {
+            mCompat = new CompatV26();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.java:33]:
```suggestion
            mCompat = new CompatV28();
```

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            public WindowInsets onApplyWindowInsets(View view, WindowInsets windowInsets) {
+                WindowInsets defaultInsets = view.onApplyWindowInsets(windowInsets);
+                final View toolbar = a.findViewById(R.id.toolbar);
+                final View answerButtons = a.findViewById(R.id.answer_options_layout);
+                final View topbar = a.findViewById(R.id.top_bar);
+                if (toolbar == null || topbar == null || answerButtons == null) {
+                    return defaultInsets;
+                }
+
+                // I think the forward-port here is to do a WindowsInsets.Builder with WindowInsets.getInsetsIgnoringVisibility
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java:87]: Commenting for a reference for later - could we go with a toast that we know it's broken for now?

File: AnkiDroid/src/test/resources/robolectric.properties [hunk 4]
```diff
@@ -0,0 +1 @@
+sdk=28
```
Comment (david-allison-1) [inline AnkiDroid/src/test/resources/robolectric.properties:1]: Can a comment be added?

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
One issue with compat - worrying, but the code looks good

**Comment by mikehardy** [comment]
Oo - that compat object return one is huge! It obviously still needs clicking around. I don't want to put anything broken out so I won't toast on the fullscreen one, I think I'd just delegate to compatv19 and log an issue

the preferences idea might just work

I'll do a polish on this the next time I have some cycles and fix everything you saw - thanks for scanning it!

**Comment by mikehardy** [comment]
superceded by #7343  and #7332

**Pull request closed without merging**